File: frontend/libs/parser/src/lib/ParsedText.ts
import { ParserRuleContext, TerminalNode } from 'antlr4';
import { Expose } from 'class-transformer';

import {
  Doc_commentContext,
  ExpressionContext,
  Field_nameContext,
} from './grammar/SheetParser';
import { Table_nameContext } from './index';
import { Span } from './Span';

export class ParsedText {
  @Expose()
  public span: Span;

  public text: string;

  constructor(span: Span, text: string) {
    this.span = span;
    this.text = text;
  }
  public static from(ctx: ParserRuleContext | null): ParsedText | undefined {
    if (!ctx) {
      return;
    }

    return new ParsedText(Span.fromParserRuleContext(ctx), ctx.getText());
  }

  public static fromTerminalNode(
    terminalNode: TerminalNode | null
  ): ParsedText | undefined {
    if (!terminalNode) {
      return;
    }

    return new ParsedText(
      Span.fromToken(terminalNode.symbol),
      terminalNode.getText()
    );
  }

  public static fromTableName(ctx: Table_nameContext): ParsedText {
    return new ParsedText(Span.fromParserRuleContext(ctx), ctx.getText());
  }

  public static fromFieldName(ctx: Field_nameContext): ParsedText {
    return new ParsedText(Span.fromParserRuleContext(ctx), ctx.getText());
  }

  public static fromExpression(
    expression: ExpressionContext | null
  ): ParsedText | undefined {
    if (!expression) {
      return;
    }

    return new ParsedText(
      Span.fromParserRuleContext(expression),
      expression.getText()
    );
  }

  public static fromDocs(ctx: Doc_commentContext[]): ParsedText[] {
    return ctx.map(
      (doc) => new ParsedText(Span.fromParserRuleContext(doc), doc.getText())
    );
  }
}

File: backend/parser/src/main/java/com/epam/deltix/quantgrid/parser/ParsedText.java
package com.epam.deltix.quantgrid.parser;

import com.epam.deltix.quantgrid.util.ParserUtils;
import com.epam.deltix.quantgrid.util.type.EscapeType;
import com.google.gson.annotations.Expose;
import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public record ParsedText(@Expose Span span, String text) {
    private static ParsedText from(ParserRuleContext ctx, String value) {
        return new ParsedText(Span.from(ctx), value);
    }

    public static ParsedText from(ParserRuleContext ctx) {
        if (ctx.exception != null) {
            return null;
        }

        return ParsedText.from(ctx, ctx.getText());
    }

    public static ParsedText from(TerminalNode terminalNode) {
        if (terminalNode == null) {
            return null;
        }

        return new ParsedText(Span.from(terminalNode.getSymbol()), terminalNode.getText());
    }

    public static ParsedText fromTableName(SheetParser.Table_nameContext ctx) {
        if (ctx == null) {
            return null;
        }

        String tableName;
        if (ctx.MULTI_WORD_TABLE_IDENTIFIER() != null) {
            tableName = stripQuotes(ctx.getText());
        } else {
            tableName = ctx.getText();
        }
        tableName = ParserUtils.decodeEscapes(tableName.trim(), EscapeType.MULTIWORD_TABLE);

        if (tableName.isEmpty()) {
            return null;
        }

        return ParsedText.from(ctx, tableName);
    }

    @Nullable
    public static ParsedText fromFieldName(SheetParser.Field_nameContext fieldNameContext) {
        if (fieldNameContext == null) {
            return null;
        }

        String fieldName = ParserUtils.decodeEscapes(stripQuotes(fieldNameContext.getText()).trim(), EscapeType.FIELD);
        if (fieldName.isEmpty()) {
            return null;
        }

        return new ParsedText(Span.from(fieldNameContext), fieldName);
    }

    public static List<ParsedText> fromDocs(List<SheetParser.Doc_commentContext> contexts) {
        List<ParsedText> result = new ArrayList<>();
        for (ParserRuleContext context : contexts) {
            ParsedText parsedText = ParsedText.from(context);
            if (parsedText != null) {
                result.add(parsedText);
            }
        }

        return result;
    }

    public static String stripQuotes(String str) {
        if (str.isEmpty()) {
            return str;
        }

        return str.substring(1, str.length() - 1);
    }
}
